pkg/ordermanager/ordermanager.go

```diff
@@ -0,0 +1,410 @@
+type Interface interface {
+	Order(ctx context.Context, domains []string, owner interface{}) (challenges []*apis.URL, cert *tls.Certificate, err error)
+}
+
+// OrderUpCallback is the signature of the function for notifying
+// owners that their order is up, and that they should invoke Order
+// again to pick it up.
+type OrderUpCallback func(owner interface{})
+
+const (
+	// Staging is the Let's Encrypt staging endpoint, which provides significantly
+	// higher quotas for use during development, but issues certificates that are
+	// not signed by it's root CA.
+	Staging = "https://acme-staging-v02.api.letsencrypt.org/directory"
+
+	// Production is the Let's Encrypt production endpoint, which issues certificates
+	// signed by their root CA.  This allows ~50 certificates to be issued per
+	// registered domain per week.
+	Production = autocert.DefaultACMEDirectory
+)
+
+var (
+	// Endpoint is the ACME API to use, it defaults to Production, but
+	// can be pointed at Staging or other compatible endpoints.
+	Endpoint = Production
+)
+
+// New creates a new OrderManager.
+func New(ctx context.Context, cb OrderUpCallback, chlr challenger.Interface) (Interface, error) {
+	acctKey, err := ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	a := &acme.Account{Contact: []string{}}
+	client := &acme.Client{
+		DirectoryURL: Endpoint,
+		UserAgent:    "knative.dev/net-http01",
+		Key:          acctKey,
+	}
+	_, err = client.Register(ctx, a, autocert.AcceptTOS)
+	if err != nil && err != acme.ErrAccountAlreadyExists {
+		return nil, err
+	}
+
+	return &impl{
+		Client:     client,
+		Callback:   cb,
+		Challenger: chlr,
+		inflight:   make(map[key]ticket, 10),
+	}, nil
+}
+
+// impl implements Interface.
+type impl struct {
+	sync.Mutex // guards access to inflight.
+
+	Challenger challenger.Interface
+	Client     *acme.Client
+	Callback   OrderUpCallback
+
+	inflight map[key]ticket
+}
+
+var _ Interface = (*impl)(nil)
+
+// ticket is used to represent an unclaimed order that is working
+// it's way through the system.
+type ticket struct {
+	uri   string
+	owner interface{}
+	err   error
+}
+
+// Order implements Interface
+func (om *impl) Order(ctx context.Context, domains []string, owner interface{}) ([]*apis.URL, *tls.Certificate, error) {
+	logger := logging.FromContext(ctx)
+	t, ok := om.getTicket(domains)
+	if !ok {
+		// If there isn't an in-flight order, then initiate a new order.
+		var err error
+		t, err = om.initiateNewOrder(ctx, domains, owner)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Fall through to return the challenges
+	}
+	if t.err != nil {
+		om.cancelOrder(ctx, domains)
+		logger.Infof("Cancelling order for %v due to error: %v", domains, t.err)
+		return nil, nil, t.err
+	}
+
+	// See if the order specified by this ticket is ready.
+	status, err := t.GetStatus(ctx, om.Client)
+	if err != nil {
+		// TODO(mattmoor): If the error isn't transient,
```

Comment:
File an issue and link?

Reply:
For some of these cases it wasn't clear to me under what circumstances the client would return the error, so the TODOs are mostly to record places where I thought there was potential for issue, but without more information I'm not sure it's actionable.